src/typedRequestData.utils.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  MessageTypeProperty,
+  MessageTypes,
+  SignTypedDataVersion,
+  TypedDataUtils,
+  TypedMessage,
+} from '@metamask/eth-sig-util';
+import type { EnvelopingTypes } from '@rsksmart/rif-relay-contracts/dist/typechain-types/contracts/RelayHub';
+
+export type EIP712Domain = TypedMessage<EnvelopingMessageTypes>['domain'];
+
+export const eip712DomainType = [
+  { name: 'name', type: 'string' },
+  { name: 'version', type: 'string' },
+  { name: 'chainId', type: 'uint256' },
+  { name: 'verifyingContract', type: 'address' },
+];
+
+export const relayDataType = [
+  { name: 'gasPrice', type: 'uint256' },
+  { name: 'feesReceiver', type: 'address' },
+  { name: 'callForwarder', type: 'address' },
+  { name: 'callVerifier', type: 'address' },
+];
+
+export const requestType = [
+  { name: 'relayHub', type: 'address' },
+  { name: 'from', type: 'address' },
+  { name: 'to', type: 'address' },
+  { name: 'tokenContract', type: 'address' },
+  { name: 'value', type: 'uint256' },
+  { name: 'nonce', type: 'uint256' },
+  { name: 'tokenAmount', type: 'uint256' },
+  { name: 'tokenGas', type: 'uint256' },
+  { name: 'data', type: 'bytes' },
+  { name: 'relayData', type: 'RelayData' },
+];
+
+export const relayRequestType = [
+  ...relayDataType,
+  { name: 'gas', type: 'uint256' },
+];
+
+export const deployRequestType = [
+  ...relayDataType,
+  { name: 'recoverer', type: 'address' },
+  { name: 'index', type: 'uint256' },
+];
+
+export type EnvelopingMessageTypes = MessageTypes & {
+  RelayRequest: MessageTypeProperty[];
+  RelayData: MessageTypeProperty[];
+};
+
+export function getDomainSeparator(
+  verifyingContract: string,
+  chainId: number
+): EIP712Domain {
+  return {
+    name: 'string name,string version',
+    version: '2',
+    chainId: chainId,
+    verifyingContract: verifyingContract,
+  };
+}
+
+export const getDomainSeparatorHash = (
+  verifier: string,
+  chainId: number
+): string =>
+  TypedDataUtils.hashStruct(
+    'EIP712Domain',
+    getDomainSeparator(verifier, chainId),
+    { EIP712Domain: eip712DomainType },
+    SignTypedDataVersion.V4
+  ).toString('hex');
+
+type GetRequestDataFieldProps = {
+  chainId: number;
+  verifier: string;
+  relayRequest:
+    | EnvelopingTypes.RelayRequestStruct
+    | EnvelopingTypes.DeployRequestStruct;
+  requestTypes: MessageTypeProperty[];
+};
+
+export const getEnvelopingRequestDataV4Field = ({
+  chainId,
+  verifier, // is this separated from the relay request by design?
```

Comment:
When would we want to use this with verifier external to the request that is being passed as third arg?

Reply:
This verifier its not verifierContracts. This verifier is the address that we are trying to recover with the domain separator. It could be either the smartWallet address, or the smartWalletFactory.

Reply:
I don't understand your answer, so I'll ask it differently:
the verifier is taken from the relay request and then the whole request is passed in as well in the account manager https://github.com/rsksmart/rif-relay-client/pull/34/files#diff-54caa818924335d964cb20a8998a3476391b16f046153c048a6147986d157e7fL92. Why is it important to keep those separate? Are there any scenarios where the verifier would not be taken from the request.

Reply:
The verifier is not taken from the relay request. This verifier is built in, in the domain separator. During the process of the signature recovering from the EIP712, the domain separator is needed to recover the external signature from the EOA from the client, not the worker.

Reply:
[see this line](https://github.com/rsksmart/rif-relay-client/blob/2d3f1668afa088d1e17e49ca7fbd515d191b93fa/src/AccountManager.ts#L92)
![Screenshot 2022-11-01 at 4 21 46 pm](https://user-images.githubusercontent.com/36888576/199284128-73cf2527-8920-42c0-8b81-ff0d6a409e92.png)
![Screenshot 2022-11-01 at 4 23 24 pm](https://user-images.githubusercontent.com/36888576/199284477-2d2c47b7-cf4c-481e-85c3-02ab8860d8b9.png)